Mixpanel/MPNetwork.m

```diff
@@ -59,26 +60,38 @@ - (void)flushQueue:(NSMutableArray *)queue endpoint:(NSString *)endpoint {
         NSString *postBody = [NSString stringWithFormat:@"ip=%d&data=%@", self.useIPAddressForGeoLocation, requestData];
         MixpanelDebug(@"%@ flushing %lu of %lu to %@: %@", self, (unsigned long)batch.count, (unsigned long)queue.count, endpoint, queue);
         NSURLRequest *request = [self requestForEndpoint:endpoint withBody:postBody];
-        NSError *error = nil;
         
         [self updateNetworkActivityIndicator:YES];
         
-        NSHTTPURLResponse *urlResponse = nil;
-        NSData *responseData = [NSURLConnection sendSynchronousRequest:request returningResponse:&urlResponse error:&error];
+        __block BOOL didFail = NO;
+        dispatch_semaphore_t semaphore = dispatch_semaphore_create(0);
+        NSURLSession *session = [NSURLSession sharedSession];
+        [[session dataTaskWithRequest:request completionHandler:^(NSData *responseData,
+                                                                  NSURLResponse *urlResponse,
+                                                                  NSError *error) {
+            [self updateNetworkActivityIndicator:NO];
+            
+            BOOL success = [self handleNetworkResponse:(NSHTTPURLResponse *)urlResponse withError:error];
+            if (error || !success) {
+                MixpanelError(@"%@ network failure: %@", self, error);
+                didFail = YES;
+            } else {
+                NSString *response = [[NSString alloc] initWithData:responseData
+                                                           encoding:NSUTF8StringEncoding];
+                if ([response intValue] == 0) {
+                    MixpanelError(@"%@ %@ api rejected some items", self, endpoint);
+                }
+            }
+            
+            dispatch_semaphore_signal(semaphore);
+        }] resume];
         
-        [self updateNetworkActivityIndicator:NO];
+        dispatch_semaphore_wait(semaphore, DISPATCH_TIME_FOREVER);
```

Comment:
Why do we need a semaphore here and not handle this inside the completion block?